File: Add Binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int m = a.length();
        int n = b.length();
        int carry = 0;
        int i, j;
        string res;
                
        for (i = m - 1, j = n - 1; i >= 0 && j >= 0; i--, j--) {
            int sum = (a[i] - '0') + (b[j] - '0') + carry;
            res.append(1, (sum%2 + '0'));
            carry = sum / 2;
        }
        
        while (i >= 0) {
            int sum = a[i] - '0' + carry;
            res.append(1, (sum%2 + '0'));
            carry = sum / 2;
            i--;
        }
        while (j >= 0) {
            int sum = b[j] - '0' + carry;
            res.append(1, (sum%2 + '0'));
            carry = sum / 2;
            j--;
        }
        if (carry == 1) res.append(1, '1');
        
        reverse(res.begin(), res.end());
        
        return res;
    }
};

File: Add Binary.cs
public class Solution {
    public string AddBinary(string a, string b) {
        if (a == null || b == null) {
            return "";
        }
        
        var na = a.Length - 1;
        var nb = b.Length - 1;
        var sb = new StringBuilder();
        int carry = 0;
        while (na >= 0 || nb >= 0) {
            var sum = carry;
            if (na >= 0) {
                sum += a[na] - '0';
                na--;
            }
            if (nb >= 0) {
                sum += b[nb] - '0';
                nb--;
            }
            carry = sum >> 1;
            sum = sum & 1;
            sb.Insert(0, sum == 0 ? '0' : '1');
        }
        if (carry > 0) {
            sb.Insert(0, '1');
        }
        
        return sb.ToString();
    }
}

